pkg/machinefilters/machine_filters.go

```diff
@@ -83,6 +86,70 @@ func MatchesKubernetesVersion(kubernetesVersion string) Func {
 // MatchesKThreesBootstrapConfig checks if machine's KThreesConfigSpec is equivalent with KCP's KThreesConfigSpec.
 func MatchesKThreesBootstrapConfig(machineConfigs map[string]*bootstrapv1.KThreesConfig, kcp *controlplanev1.KThreesControlPlane) Func {
 	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+
+		// Check if KCP and machine KThreesServerConfig matches, if not return
+		if !matchKThreesServerConfig(kcp, machine) {
+			return false
+		}
+
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			// This is a safety precaution to avoid selecting machines that are broken, which in the future should be remediated separately.
+			return true
+		}
+
+		machineConfig, found := machineConfigs[machine.Name]
+		if !found {
+			// Return true here because failing to get KThreesConfig should not be considered as unmatching.
+			// This is a safety precaution to avoid rolling out machines if the client or the api-server is misbehaving.
+			return true
+		}
+
+		kcpConfig := kcp.Spec.KThreesConfigSpec.DeepCopy()
+		// KCP bootstrapv1.KThreesServerConfig will only be compared with a machine's ServerConfig annotation, so
+		// we are cleaning up from the reflect.DeepEqual comparison.
+		kcpConfig.ServerConfig = bootstrapv1.KThreesServerConfig{}
+		machineConfig.Spec.ServerConfig = bootstrapv1.KThreesServerConfig{}
+		// KCP version check is handled elsewhere
+		kcpConfig.Version = ""
+		machineConfig.Spec.Version = ""
+
+		return reflect.DeepEqual(&machineConfig.Spec, kcpConfig)
+	}
+}
+
+// matchKThreesServerConfig verifies if KCP and machine ClusterConfiguration matches.
+// NOTE: Machines that have KThreesServerConfigurationAnnotation will have to match with KCP KThreesServerConfig.
+// If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes
+// made in KCP's KThreesServerConfig given that we don't have enough information to make a decision.
+// Users should use KCP.Spec.RolloutAfter field to force a rollout in this case.
+func matchKThreesServerConfig(kcp *controlplanev1.KThreesControlPlane, machine *clusterv1.Machine) bool {
```

Comment:
at least for controlplane, init and join cp node has same bootstrap config, this is different from how kubeadm works
/home/seed/repo/cluster-api-k3s/pkg/k3s/control_plane.go

I'm not very sure about worker node, you may provision a k3s cluster and check the bootstrap config.

Reply:
Yeah I removed the init and join config sections. The bits I copied were:
- Check the serverConfig via annotation on the machine
- Check other controlplane config against the machine spec via the machineConfigs map
Open to suggestions on how to simply though?

Reply:
Ah, I see what you mean. The init and join config is just the kcp KThreesConfigSpec, we do still compare this [here](https://github.com/k3s-io/cluster-api-k3s/pull/114/files#diff-956e3de66d9eb65a63c655c93cec23422aafab713ed5c9343dd9159322cbba66R154), since the init and join config are just the kcp KThreesConfigSepc, how do you suggest we should simplify the code?

Reply:
i'm not sure if worker node's bootstrap config is the same with kcp config, can you confirm?
if yes, then we can simply compare kcp config and machine bootstrap config directly, no need to compare serverConfig separately

Reply:
Addressed in the latest commit